### PR TITLE
Cache evaluated directories in isExcludedByFile

### LINKED ISSUE #1271: --exclude-if-present leads to lots of lstat-operations for each directory
## Output of `restic version`

restic 0.7.3 (v0.7.3-28-g2080afd-dirty)
compiled with go1.9 on linux/amd64

## How did you run restic exactly?

export RESTIC_REPOSITORY=/srv/storage/restic
export RESTIC_PASSWORD_FILE=$HOME/.config/restic-password

ionice -c 3 ./restic backup --exclude-caches --exclude /mnt/backup-snapshot/_backup --force /mnt/backup-snapshot

## What backend/server/service did you use?
local mount of ZFS via CIFS

## Expected behavior
lstat each CACHEDIR.TAG once per directory

## Actual behavior
CACHEDIR.TAG is lstat'ed once per file 

```
futex(0xc4204cf610, FUTEX_WAKE, 1)      = 1 <0.000007>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/7969-tiny-816f5f5ad4039bdb6eb0bdeebe3fe69ed4f83c3a", {st_mode=S_IFREG|0600, st_size=7425, ...}) = 0 <0.000024>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/CACHEDIR.TAG", 0xc545816518) = -1 ENOENT (No such file or directory) <0.000008>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/7969-tiny-816f5f5ad4039bdb6eb0bdeebe3fe69ed4f83c3a", {st_mode=S_IFREG|0600, st_size=7425, ...}) = 0 <0.000010>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/CACHEDIR.TAG", 0xc5458166b8) = -1 ENOENT (No such file or directory) <0.000007>
epoll_wait(4, {}, 128, 0)               = 0 <0.000006>
futex(0xe3feb0, FUTEX_WAIT, 0, NULL)    = 0 <0.008781>
epoll_wait(4, {}, 128, 0)               = 0 <0.000010>
pselect6(0, NULL, NULL, NULL, {0, 3000}, 0) = 0 (Timeout) <0.000064>
futex(0xc420108490, FUTEX_WAKE, 1)      = 1 <0.000013>
futex(0xc420108490, FUTEX_WAKE, 1)      = 1 <0.000008>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/8236-81177c7d3b2c70e6a9a60bc53bdbc29b5a857be8", {st_mode=S_IFREG|0600, st_size=587571, ...}) = 0 <0.000068>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/CACHEDIR.TAG", 0xc5fbc7f628) = -1 ENOENT (No such file or directory) <0.000009>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/8236-81177c7d3b2c70e6a9a60bc53bdbc29b5a857be8", {st_mode=S_IFREG|0600, st_size=587571, ...}) = 0 <0.000007>
lstat("/mnt/backup-snapshot/[..sensitive pathname..]/cache/81/CACHEDIR.TAG", 0xc5fbc7f7c8) = -1 ENOENT (No such file or directory) <0.000008>
```

## Steps to reproduce the behavior

run restic with --exclude-caches or own --exclude-if-present

## Do you have any idea what may have caused this?
From looking at the code I'd judge that the problem is, that the exclude filter is only applied to entries of each path that is walked in Pipe.Walk. If it would be applied to the path itself, too, and skip walking the entries if the path itself is excluded, this should not happen. Otherwise this seems not the complete solution, as if it were true, the problem should only happen in the first level of path traversal and not in "deeper" levels. So I'd say my analysis is flawed somehow.

## Do you have an idea how to solve the issue?

I see multiple possible solutions:

1. split exclude filters for paths and entries. This seems to be the most effective solution: if the exclude filters are split (--exclude-if-present goes only with path and not with entry filters) then there is no need to check the presence of these files for each entry, but rather for each path
2. implement a temporary cache for the files presence(s)

### PR DESCRIPTION
**Pull request by fawick**
Fixes #1271.

### CODE REVIEW

**Comment by fawick** [comment]
@phillipp I'd be interested to learn whether this PR improves your usecase from #1271

File: cmd/restic/exclude.go [hunk 1]
```diff
@@ -60,22 +68,43 @@ func rejectIfPresent(excludeFileSpec string) (RejectFunc, error) {
...
+// directory based on previous visits.
+func isExcludedByFile(filename, tagFilename, header string, rc *rejectionCache) bool {
 	if tagFilename == "" {
 		return false
 	}
 	dir, base := filepath.Split(filename)
 	if base == tagFilename {
 		return false // do not exclude the tagfile itself
 	}
+	if rc != nil {
```
Comment (fd0) [inline cmd/restic/exclude.go]: It'd be nice if this were a method on the `rc` struct, and it should also work if `rc` is nil, so we can abstract that away, like `rejected, ok := rc.Get(dir); if ok { return rejected }`. What do you think?

File: cmd/restic/exclude.go [hunk 2]
```diff
@@ -60,22 +68,43 @@ func rejectIfPresent(excludeFileSpec string) (RejectFunc, error) {
...
+		rc.mtx.Lock()
+		defer rc.mtx.Unlock()
+		if rejected, visited := rc.m[dir]; visited {
+			return rejected
+		}
+	}
+
+	rejected := isDirExcludedByFile(dir, tagFilename, header)
+	if rc != nil {
```
Comment (fd0) [inline cmd/restic/exclude.go]: This one too, e.g. `rc.Store(dir, rejected)`.

**Review by fd0 — CHANGES_REQUESTED** [review]

File: cmd/restic/exclude.go (hunk 1 above)
Comment (fawick) [reply cmd/restic/exclude.go]: Good idea. The mutex must remain locked until the function returnes, though, so I wrap its `Lock()` and `Unlock()` methods, too.

File: cmd/restic/exclude.go (hunk 2 above)
Comment (fawick) [reply cmd/restic/exclude.go]: Done.

**Review by fawick — COMMENTED** [review]
@fd0 The changes you requested have been implemented.

**Review by fd0 — APPROVED** [review]

**Pull request merged**